### ISSUE TITLE
API: integrate with openpyxl 2.0-2.1 changes

### ISSUE THREAD

**Issue by jtratner** [pydata/pandas#8342]
Builds are failing on Travis due to some issue with Excel (weirdly passed on merge but then started failing a few commits later). Currently investigating. C.f. https://travis-ci.org/pydata/pandas/jobs/35873580

@neirbowj - any idea why these tests stopped passing?

It's all some variation on the following:

```
======================================================================
ERROR: test_tsframe (pandas.io.tests.test_excel.OpenpyxlTests_NoMerge)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/io/tests/test_excel.py", line 562, in test_tsframe
    df.to_excel(path, 'test1')
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/util/decorators.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/core/frame.py", line 1228, in to_excel
    startrow=startrow, startcol=startcol)
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/io/excel.py", line 687, in write_cells
    'number_format':{'format_code': self.datetime_format}}))
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/io/excel.py", line 757, in _convert_to_style_kwargs
    new_v = _conv_to_x(v)
  File "/home/travis/virtualenv/python3.4.1/lib/python3.4/site-packages/pandas/io/excel.py", line 1043, in _convert_to_number_format
    from openpyxl.styles import NumberFormat
ImportError: cannot import name 'NumberFormat'
----------------------------------------------------------------------
```

**Comment by stancikcom** [comment]
Guys, IO modules are essential for Pandas usability. I falled into dependency hell as another module (PETL) requires newer version of Openpyxl. Please, consider to implement an adapter pattern that will support (reasonable legacy versions). 

I think Pandas is a great package and for production usage it is worth to invest more into reliability of the essential modules.

**Comment by anon3** [comment]
Hi, I am trying to create an Excel File(showing the differences of two Excel files). I am getting this error:TypeError: expected <class 'openpyxl.styles.fonts.Font'>.
I have already imported : >>> import openpyxl.styles
>>> from openpyxl.styles import fonts

**Comment by jreback** [comment]
@stancikcom and @anon3 
pandas has been compat with all released versions of openpyxl for as far back as I can remember 
there were a couple of hiccups iirc 0.15.0
0.15.2 has been out for quite some time and 0.16.0 just released
if u are having difficulties thrn pls pd.show_versions() in your report

**Comment by anon3** [comment]
@jreback The version is pandas: 0.15.2. Would you suggest me to upgrade it to 0.16.0 ?

**Comment by jreback** [comment]
@anon3 well pls report ``pd.show_versions()`` as well as as showing what is causing the error

**Comment by stancikcom** [comment]
Hi all,

after upgrade, pandas works fine with new version of openpyxl.
my `import pandas; pandas.show_versions()` dump:

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-46-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.0
nose: 1.3.4
Cython: 0.21.2
numpy: 1.9.2
scipy: 0.15.0
statsmodels: 0.6.1
IPython: 2.3.1
sphinx: 1.2.3
patsy: 0.2.1
dateutil: 2.4.1
pytz: 2015.2
bottleneck: 1.0.0
tables: 3.1.1
numexpr: 2.2.2
matplotlib: 1.4.2
openpyxl: 2.2.0
xlrd: 0.9.2
xlwt: 0.7.5
xlsxwriter: 0.6.7
lxml: 3.3.3
bs4: 4.2.1
html5lib: 0.999
httplib2: 0.8
apiclient: None
sqlalchemy: 0.9.9
pymysql: 0.6.6.None
psycopg2: 2.6 (dt dec pq3 ext lo64)

**Comment by anon3** [comment]
@jreback this is what I used:
>>> import pandas as pd
>>> pd.show_versions()
INSTALLED VERSIONS
------------------
commit: None
python: 3.4.2.final.0
python-bits: 32
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 42 Stepping 7, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.16.0
nose: None
Cython: None
numpy: 1.9.2
scipy: None
statsmodels: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.4.0
pytz: 2014.10
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.4.3
openpyxl: 2.2.0
xlrd: 0.9.3
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None

**Comment by anon3** [comment]
@jreback  and @stancikcom even with 0.16.0 i get this error: TypeError: expected <class 'openpyxl.styles.fonts.Font'>

**Comment by stancikcom** [comment]
@anon3 try to share a snippet of your code which makes you a trouble. We can try to replicate / debug it. Cheers. (ms)

**Comment by anon3** [comment]
@stancikcom : here is the code that I found (and  I am using for official purpose) to compare two excels:

import pandas as pd
import numpy as np
import xlrd

# Define the diff function to show the changes in each field
def report_diff(x):
    return x[0] if x[0] == x[1] else '{} ---> {}'.format(*x)

# We want to be able to easily tell which rows have changes
def has_change(row):
    if "--->" in row.to_string():
        return "Y"
    else:
        return "N"

# Read in both excel files
df1 = pd.read_excel('Excel1.xlsx', 'Sheet2', na_values=['NA'])
df2 = pd.read_excel('Excel2.xlsx', 'Sheet2', na_values=['NA'])

# Make sure we order by account number so the comparisons work
df1.sort(columns="Col1")
df1=df1.reindex()
df2.sort(columns="Col1")
df2=df2.reindex()

# Create a panel of the two dataframes
diff_panel = pd.Panel(dict(df1=df1,df2=df2))

#Apply the diff function
diff_output = diff_panel.apply(report_diff, axis=0)

# Flag all the changes
diff_output['has_change'] = diff_output.apply(has_change, axis=1)

#Save the changes to excel but only include the columns we care about
diff_output[(diff_output.has_change == 'Y')].to_excel('my-diff-1.xlsx',index=False,columns=["Col1",
                                                      "Col2","Col3","Col4","Col5"])

**Comment by stancikcom** [comment]
@anon3  I have tried your code on my data sets (of course I do have different data than you) and  I can confirme that the pure IO from/to excel worked without any problem in my python environment. 

However, I think the issues are in your code logic. Currently it won't work, if the data sets are  different size, it won't work if the `account` order is altered e.g. due to added/deleted `account` in one of the set. It sound to be an overkill to be using 3D `pandas.Panel()` ;-)

If I were you I would refactor the logic  as follows:
-assuming `account` is a unique primary key for each row, then even without Pandas you can write your function using built_in `hasattr()` function. `Accounts` could be a hash key and its value the account attributes (a tuple where each member is taken from one cell from excel). So then
``` python 
data_set = { Account_number_i : (Col1,Col2,Col3,...Coln), ... } 
```

With Pandas the logic will be as follows:
1. use pandas IO function to extract data from excel
```python
df1 = pd.read_excel('Excel1.xlsx', 'Sheet2', na_values=['NA'])
df2 = pd.read_excel('Excel2.xlsx', 'Sheet2', na_values=['NA'])
```

2. merge both datasets, 
```python
dfm = pd.merge(df1, df2, on='account', how='left')
```
if a row in the `dfm` has `np.NaN` values that means that `account` has been added/deleted   

3. iterate over rows where  `account` exists in the both sets, check if the attributes are the same (the tuples has to be equal), else compare each tuple member to track changes.  
```python
# try it to write this yourself, if you still struggle, let me know.
```
4. use pandas IO function to make final export to excel...
```python
# try it to write this yourself, if you still struggle, let me know.
```
Hopefuly it helps.
cheers.

**Comment by anon3** [comment]
@stancikcom  thanks for your valuable comments. I am a newbie .. never used hasattr() function.
Let me try learning and using hasattr()

**Comment by stancikcom** [comment]
@anon3  with #python and #pandas you are in the right track... :-) 
bdw. if you want to still solve your problem with pandas, merge `df1` and `df2` on the `account` column, and then write short function to compare 'left'  and  'right' side of the table. 
```python
pd.merge(df1,df2,on='account',how='left')
```

**Comment by anon3** [comment]
@stancikcom ,
Yes. I wrote something something and I get : ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

This is the additional piece of code:
while(df1 and df2):    
    if(df1[0:len(df1)]==df2[0:len(df2)]):
        continue
    else:
        # something

PS:  I am sure there is something wrong in comparison.I'll try to use either of these:a.empty, a.bool(), a.item(), a.any() or a.all()
I'll keep you updated with my final code

**Comment by anon3** [comment]
@stancikcom ,

how to compare the left and right column from the merged dataframe?

I was trying to find some reference over the stackflow, didn't find for the merge but found a different way to compare two data frames. Here is the piece of code which works fine:
import pandas as pd
import numpy as np
import xlrd

# Define the diff function to show the changes in each field
def report_diff(x):
    return x[0] if x[0] == x[1] else '{} ---> {}'.format(*x)

# We want to be able to easily tell which rows have changes
def has_change(row):
    if "--->" in row.to_string():
        return "Y"
    else:
        return "N"

# Read in both excel files
df1 = pd.read_excel('Excel1.xlsx', 'Sheet2', na_values=['NA'])
df2 = pd.read_excel('Excel2.xlsx', 'Sheet2', na_values=['NA'])

# Make sure we order by account number so the comparisons work
df1.sort(columns="Col1")
df1=df1.reindex()
df2.sort(columns="Col1")
df2=df2.reindex()

ne_stacked = (df1 != df2).stack()

changed = ne_stacked[ne_stacked]

changed.index.names = ['id', 'col']



difference_locations = np.where(df1 != df2)

changed_from = df1.values[difference_locations]

changed_to = df2.values[difference_locations]

print(pd.DataFrame({'from': changed_from, 'to': changed_to}, index=changed.index))

**Comment by jreback** [comment]
closed in favor of #10125
